src/oc_telemetry.erl

```diff
@@ -37,63 +55,106 @@ attach(Name, EventName, Description, Unit) ->
     Measure = oc_stat_measure:new(Name, Description, Unit),
-    Options = #options{name = Name, filter = Metadata},
-
-    case telemetry:attach(Name, EventName, fun handle_event/4, Options) of
-      ok -> {ok, Measure};
-      Error -> Error
+    OCMeasurement = oc_measurement(Name, Measurement, TagValues),
+    case telemetry:attach(Name, EventName,
+                          fun handle_event/4, OCMeasurement) of
+        ok ->
+            {ok, Measure};
+        Error ->
+            Error
     end.
 
-% Handle events send by `telemetry' application
-handle_event(_EventName, Value, Metadata,
-             #options{name = Name, filter = Filter}) ->
-  Filtered = Filter(Metadata),
-
-  ok = oc_stat:record(Filtered, Name, Value).
+oc_measurement(Name, Measurement, TagValues) when is_atom(Measurement) ->
+    #oc_measurement{name=Name,
+                    value_fun=fun(Values) ->
+                                  maps:get(Measurement, Values, undefined)
+                              end,
+                    tag_values=TagValues};
+oc_measurement(Name, Measurement, TagValues) when is_function(Measurement) ->
+    #oc_measurement{name=Name,
+                    value_fun=Measurement,
+                    tag_values=TagValues}.
+
+%% Handle events send by `telemetry' application
+handle_event(_EventName, Values, Tags, #oc_measurement{name=Name,
+                                                       value_fun=ValueFun,
+                                                       tag_values=TagValues}) ->
+    case ValueFun(Values) of
+        undefined ->
+            ok;
+        Value ->
+            ok = oc_stat:record(TagValues(Tags), Name, Value)
+    end;
+handle_event(_EventName, Values, Tags, Measurements) ->
+    lists:foreach(fun({Key, Name}) ->
+                          case maps:get(Key, Values, undefined) of
+                              undefined ->
+                                  ok;
+                              Value ->
+                                  ok = oc_stat:record(Tags, Name, Value)
+                          end
+                  end, Measurements).
 
 build_name(NormalizedName) ->
-  Stringified = [atom_to_list(Atom) || Atom <- NormalizedName],
-  Joined = lists:join($/, Stringified),
-
-  list_to_binary(Joined).
+    Stringified = [atom_to_list(Atom) || Atom <- NormalizedName],
+    Joined = lists:join($/, Stringified),
+    list_to_binary(Joined).
 
 aggregation('Elixir.Telemetry.Metrics.Counter', _) ->
-  oc_stat_aggregation_count;
+    oc_stat_aggregation_count;
 aggregation('Elixir.Telemetry.Metrics.Sum', _) ->
-  oc_stat_aggregation_sum;
+    oc_stat_aggregation_sum;
 aggregation('Elixir.Telemetry.Metrics.LastValue', _) ->
-  oc_stat_aggregation_latest;
+    oc_stat_aggregation_latest;
 aggregation('Elixir.Telemetry.Metrics.Distribution', #{buckets := Buckets}) ->
-  {oc_stat_aggregation_distribution, [{buckets, Buckets}]}.
+    {oc_stat_aggregation_distribution, [{buckets, Buckets}]}.
 
```

Comment:
Might want to handle the new Summary type being passed even if to just log that it's unsupported and continue on.